File: snippets/SortListOfSortedLists.cpp
#include <set>
#include <vector>
#include <string>
#include <iostream>

using namespace std;

template <typename T>
class list_item {
	public:
		list_item(T item, int index, int position) : item { item }, index { index }, position { position } {}

		T getItem() const {
			return item;
		}

		int getIndex() const {
			return index;
		}

		int getPosition() const {
			return position;
		}

		bool operator< (list_item<T> other) const {
			if (getItem() != other.getItem()) {
				return getItem() < other.getItem();
			}
			return getIndex() < other.getIndex();
		}

	private:
		T item;
		int index, position;
};

template <typename T>
vector<T>* sort(vector<vector<T>>& lists) {
	set<list_item<T>> mins;
	for (unsigned int j = 0; j < lists.size(); j++) {
		if (lists[j].size() > 0) {
			mins.emplace(list_item<T>(lists[j][0], j, 0));
		}
	}
	
	vector<T>* result = new vector<T>();

	while (mins.size() > 0) {
		auto min = mins.begin();
		result->push_back(min->getItem());
		if (lists[min->getIndex()].size() > min->getPosition() + 1) {
			mins.emplace(list_item<T>(lists[min->getIndex()][min->getPosition() + 1], min->getIndex(), min->getPosition() + 1));
		}
		mins.erase(min);
	}

	return result;
}

int main() {
	vector<vector<string>> lists { { "1", "2", "3", "7" }, { "1", "5", "6" } };
	vector<string>* sorted = sort(lists);
	for (auto& item : *sorted) {
		cout << item.c_str() << endl;
	}
	delete sorted;
	return 0;
}

File: snippets/SortListOfSortedLists.java
import java.util.ArrayList;
import java.util.Arrays;
import java.util.List;
import java.util.TreeSet;

class ListItem<T extends Comparable<T>> implements Comparable<ListItem<T>> {
	private T item;
	private int index;
	private int position;

	public ListItem(T item, int index, int position) {
		this.item = item;
		this.index = index;
		this.position = position;
	}
	
	public T getItem() {
		return item;
	}
	
	public int getIndex() {
		return index;
	}
	
	public int getPosition() {
		return position;
	}
	
	@Override
	public int compareTo(ListItem<T> o) {
		if (getItem().compareTo(o.getItem()) != 0) {
			return getItem().compareTo(o.getItem());
		}
		return Integer.compare(getIndex(), o.getIndex());
    }
}

public class SortListOfSortedLists {

	public static void main(String[] args) {
		ArrayList<List<String>> input = new ArrayList<List<String>>();
		input.add(Arrays.asList(new String[] { "A", "E", "Z" }));
		input.add(Arrays.asList(new String[] { "A", "D", "X" }));
		List<String> result = sort(input);
		for (String s : result) {
			System.out.println(s);
		}
	}

	private static <T extends Comparable<T>> List<T> sort(ArrayList<List<T>> input) {
		TreeSet<ListItem<T>> mins = new TreeSet<ListItem<T>>();
		ArrayList<T> result = new ArrayList<T>();
		
		for (int j = 0; j < input.size(); j++) {
			List<T> l = input.get(j);
			if (l.size() > 0) {
				mins.add(new ListItem<T>(l.get(0), j, 0));
			}
		}
		
		while (mins.size() > 0) {
			ListItem<T> min = mins.pollFirst();
			result.add(min.getItem());
			if (input.get(min.getIndex()).size() > min.getPosition() + 1) {
				mins.add(new ListItem<T>(input.get(min.getIndex()).get(min.getPosition() + 1), min.getIndex(), min.getPosition() + 1));
			}
		}
		
		return result;
    }
    
}
